src/xma/src/xmaplugin/xmaplugin.cpp

```diff
@@ -550,37 +577,38 @@ XmaCUCmdObj xma_plg_schedule_work_item(XmaSession s_handle,
     
     bool expected = false;
     bool desired = true;
-    while (!priv1->execbo_locked.compare_exchange_weak(expected, desired)) {
-        //std::this_thread::sleep_for(std::chrono::milliseconds(1));
-        expected = false;
-    }
-    //kernel completion lock acquired
-    //xma_logmsg(XMA_DEBUG_LOG, XMAPLUGIN_MOD, "1. Num of cmds in-progress = %lu\n", priv1->CU_cmds.size());
+    int32_t bo_idx = -1;
 
     // Find an available execBO buffer
     uint32_t itr = 0;
-    int32_t bo_idx = -1;
     while (true) {
-        bo_idx = xma_plg_execbo_avail_get(s_handle);
+        expected = false;
+        while (!priv1->execbo_locked.compare_exchange_weak(expected, desired)) {
+            std::this_thread::yield();
+            expected = false;
+        }
+
+        if (g_xma_singleton->cpu_mode == XMA_CPU_MODE2) {
+            bo_idx = xma_plg_execbo_avail_get2(s_handle);
+        } else {
+            bo_idx = xma_plg_execbo_avail_get(s_handle);
+        }
         if (bo_idx != -1) {
             break;
         }
+        xma_logmsg(XMA_DEBUG_LOG, XMAPLUGIN_MOD, "No available execbo found");
+        priv1->execbo_locked = false;
         if (itr > 15) {
-            xma_logmsg(XMA_DEBUG_LOG, XMAPLUGIN_MOD, "Unable to find free execbo to use\n");
-            priv1->execbo_locked = false;
+            xma_logmsg(XMA_ERROR_LOG, XMAPLUGIN_MOD, "Unable to find free execbo to use\n");
             if (return_code) *return_code = XMA_ERROR;
             return cmd_obj_error;
         }
-        std::this_thread::sleep_for(std::chrono::milliseconds(3));
+        std::unique_lock<std::mutex> lk(priv1->m_mutex);
+        priv1->execbo_is_free.wait(lk);
+        lk.unlock();
```

Comment:
Are you ok with spurious wake ups here?  Normally wait() is used with a some condition which is checked at wake-up:
https://en.cppreference.com/w/cpp/thread/condition_variable/wait

No point calling lk.unlock(). The lock guard does that automatically at scope exit which is one line later.

Reply:
Based on comments that this has many change. I will setup separate review of couple of file at a time.